feat: add useSearchParams

wd-David opened the pull request:
We used this utility function internally for 2 months, and I think it's stable enough to create a PR here.

I mentioned this to @TGlide during the Svelte Summit, and here are the things we need to discuss before proceeding.
- Inspired by [sveltekit-search-params](https://github.com/paoloricciuti/sveltekit-search-params)(@paoloricciuti you rock!), some options are similar. However, I chose the [Standard Schema](https://standardschema.dev/) instead of providing encoding/ decoding capabilities.
- `useSearchParams` uses SvelteKit primitives, so this will make the `runed` library not just for Svelte. Thomas mentioned we can probably use something like `runed/kit` and check if SvelteKit exists in the peer dependencies.
- I use [`dequal`](https://github.com/lukeed/dequal) for variable comparision and [`lz-string`](https://github.com/pieroxy/lz-string?tab=readme-ov-file) for compression, so the bundle size will increase.
- I added E2E tests (and updated `ci.yaml`) since the original vitest browser mode setup is insufficient. The CI time will also increase.

Any feedback or suggestions are welcome.

paoloricciuti commented:
This is great...I'll take a look as soon as I have time.

paoloricciuti commented on packages/runed/src/lib/utilities/index.ts:
```diff
@@ -26,3 +26,4 @@ export * from "./use-resize-observer/index.js";
 export * from "./watch/index.js";
 
 export * from "./scroll-state/index.js";
+export * from "./use-search-params/index.js";
```
I don't think we can do this...we are importing sveltekit stuff in the module that will fail if you try to use runed in a svelte only project.

TGlide replied:
yeah, we should have a /kit sub-import path `import * from 'runed/kit'`

TGlide commented:
Also, thank you so much!

wd-David commented on packages/runed/vite.config.ts:
```diff
@@ -26,8 +26,8 @@ export default defineConfig({
 					setupFiles: ["./setupTest.ts"],
-					include: ["src/**/*.{test,test.svelte,spec}.{js,ts}"],
-					exclude: ["src/**/*.browser.{test,test.svelte,spec}.{js,ts}"],
+					include: ["src/**/*.{test,test.svelte}.{js,ts}"],
+					exclude: ["src/**/*.browser.{test,test.svelte}.{js,ts}"],
```
To separate testing files, vitest now only finds `test.{js,ts}`, and playwright finds `spec.{js,ts}`.

wd-David commented:
@paoloricciuti @TGlide I updated the export path, and it should work as expected now. 
This PR is ready for review!

TGlide commented:
Sorry for the massive wait David! Taking another look

TGlide commented:
Hey @wd-David I pushed a demo, but it does not work as expected for me. Also added a comment in the demo file. LMK what you think!

Overall I love the API though, thank you *so* much.

wd-David commented:
> Hey @wd-David I pushed a demo, but it does not work as expected for me. Also added a comment in the demo file. LMK what you think!
> 
> Overall I love the API though, thank you _so_ much.

@TGlide Sorry for the late reply, and thank you for the demo.
The current design only provides top-level reactivity. I added a note in the doc about this.

PR is now ready to review & merge.

wd-David commented:
> Hey @wd-David, awesome looking API! Does this not support SSR?
> 
>  CleanShot.2025-08-26.at.16.40.04.mp4

It fully supports SSR.  
The behavior shown in the clip is expected—those values are initialized and validated after hydration, since URLSearchParams are only available in the browser.

huntabyte reviewed (approved):
Awesome job, looking forward to using this in some projects!
